drivers/gpio/gpio_adp5585.c

```diff
@@ -0,0 +1,478 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/drivers/mfd/adp5585.h>
+
+#define DT_DRV_COMPAT adi_adp5585_gpio
+
+LOG_MODULE_REGISTER(adp5585_gpio, CONFIG_GPIO_LOG_LEVEL);
+
+#define ADP5585_BANK(offs) (offs >> 3)
+#define ADP5585_BIT(offs)  (offs & 0b0111)
+
+enum adp5585_gpio_pin_direction {
+	adp5585_pin_input = 0U,
+	adp5585_pin_output,
+};
+
+enum adp5585_gpio_pin_drive_mode {
+	adp5585_pin_drive_pp = 0U,
+	adp5585_pin_drive_od,
+};
+
+enum adp5585_gpio_pull_config {
+	adp5585_pull_up_300k = 0U,
+	adp5585_pull_dn_300k,
+	adp5585_pull_up_100k, /* not used */
+	adp5585_pull_disable,
+};
+
+enum adp5585_gpio_int_en {
+	adp5585_int_disable = 0U,
+	adp5585_int_enable,
+};
+
+enum adp5585_gpio_int_level {
+	adp5585_int_active_low = 0U,
+	adp5585_int_active_high,
+};
+
+/** Configuration data */
+struct adp5585_gpio_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	const struct device *mfd_dev;
+	const struct gpio_dt_spec gpio_int;
+};
+
+/** Runtime driver data */
+struct adp5585_gpio_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	uint16_t output;
+
+	sys_slist_t callbacks;
+};
+
+static int gpio_adp5585_config(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	const struct adp5585_gpio_config *cfg = dev->config;
+	struct adp5585_gpio_data *data = dev->data;
+	const struct mfd_adp5585_config *parent_cfg =
+		(struct mfd_adp5585_config *)(cfg->mfd_dev->config);
+	struct mfd_adp5585_data *parent_data = (struct mfd_adp5585_data *)(cfg->mfd_dev->data);
+
+	int ret = 0;
+	uint8_t reg_value;
+
+	/* ADP5585 has non-contiguous gpio pin layouts, account for this */
+	if ((pin & cfg->common.port_pin_mask) == 0) {
+		LOG_ERR("pin %d is invalid for this device", pin);
+		return -ENOTSUP;
+	}
+
+	uint8_t bank = ADP5585_BANK(pin);
+	uint8_t bank_pin = ADP5585_BIT(pin);
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Single Ended lines (Open drain and open source) not supported */
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* Simultaneous PU & PD mode not supported */
+	if (((flags & GPIO_PULL_UP) != 0) && ((flags & GPIO_PULL_DOWN) != 0)) {
+		return -ENOTSUP;
+	}
+
+	/* Simultaneous input & output mode not supported */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&parent_data->lock, K_FOREVER);
+
+	if ((flags & GPIO_SINGLE_ENDED) != 0) {
```

Comment:
This condition already causes function exit at line 93

Reply:
Oh, line 93 was a mistate... thanks for pointing out.